File: ports/jvm/src/main/java/com/unicodesecurity/SkinToneVariationForgery.java
package com.unicodesecurity;

import java.util.ArrayList;
import java.util.List;

/**
 * skin-tone-variation-forgery — identity-layer detector for skin-tone modifier
 * and variation-selector abuse on emoji bases per UTS #51. Answers the question:
 * does this codepoint sequence place a skin-tone modifier on a codepoint that
 * cannot bear one, stack multiple skin tones on a single base, or force a
 * text-style render on an emoji-default codepoint?
 *
 * <p>Direct port of {@code Unicode/Security/Identity/SkinToneVariationForgery.lean},
 * transliterated byte-faithfully from the verified Rust reference implementation.
 *
 * <p>Threat model. An adversary places a skin-tone modifier on a codepoint that
 * does NOT bear {@code Emoji_Modifier_Base}, stacks multiple skin tones on one
 * base, or forces a text-style render on an emoji-default codepoint via
 * {@code U+FE0E} (VS15) — sometimes to hide a payload-bearing glyph in plain
 * sight. Distinct from a pair-aligned variation-selector payload run that decodes
 * to bytes: this catches the orthogonal case of <em>semantic</em> variation-selector
 * / skin-tone misuse on a single base.
 *
 * <p>Emoji property data. The skin-tone modifier predicate reuses the port's own
 * {@link EmojiZwjIntegrity#isEmojiModifier} (the {@code U+1F3FB..U+1F3FF} set).
 * The {@code Emoji_Modifier_Base} and {@code Emoji_Presentation} property rows are
 * parsed from the port's already-bundled SHA-pinned {@code data/emoji-data.txt}
 * via the port's own {@link Security#readResource} loader (the same file the
 * {@link AiWatermarkDetectability} probe reads its {@code Emoji} rows from), never
 * a host emoji library and never String normalization.
 *
 * <p>Sub-threats (priority order, first match wins):
 * <ol>
 *   <li>{@code StackedSkinTones}      — a base immediately followed by &ge; 2 skin-tone modifiers.</li>
 *   <li>{@code InvalidSkinToneTarget} — a skin-tone modifier on a non-{@code Emoji_Modifier_Base}.</li>
 *   <li>{@code ForcedTextStyle}       — {@code U+FE0E} on an {@code Emoji_Presentation} codepoint.</li>
 * </ol>
 */
public final class SkinToneVariationForgery {
  private SkinToneVariationForgery() {}

  /** UTS #39 family key; the reason-code layer for this family is {@code I}. */
  public static final String FAMILY = "skin-tone-variation-forgery";

  /** The text-style variation selector VS15. */
  public static final int VS15 = 0xFE0E;

  /** The emoji-style variation selector VS16. */
  public static final int VS16 = 0xFE0F;

  // ───────────────────────────────────────────────────────────────────────
  // §1 Types
  // ───────────────────────────────────────────────────────────────────────

  /** Sub-threats this detector can fire, in priority order. */
  public sealed interface SubThreat
      permits StackedSkinTones, InvalidSkinToneTarget, ForcedTextStyle {
    /** Fixture-row tag string for this sub-threat (matches {@code SubThreat.tag}). */
    String tag();
  }

  /**
   * A base at {@code basePos} followed by &ge; 2 skin-tone modifiers.
   * {@code modifiers} carries the first two stacked skin-tone modifiers.
   */
  public record StackedSkinTones(int basePos, List<Integer> modifiers) implements SubThreat {
    @Override
    public String tag() {
      return "StackedSkinTones";
    }
  }

  /**
   * A skin-tone {@code modifierCp} at {@code basePos + 1} on a non-modifier-base
   * {@code baseCp}.
   */
  public record InvalidSkinToneTarget(int basePos, int baseCp, int modifierCp) implements SubThreat {
    @Override
    public String tag() {
      return "InvalidSkinToneTarget";
    }
  }

  /**
   * A {@code U+FE0E} at {@code basePos + 1} forcing text style on an
   * {@code Emoji_Presentation} {@code baseCp}.
   */
  public record ForcedTextStyle(int basePos, int baseCp) implements SubThreat {
    @Override
    public String tag() {
      return "ForcedTextStyle";
    }
  }

  /** Top-level classification for SkinToneVariationForgery. */
  public sealed interface Classification permits Clear, Hazard {
    /** True iff the classification is {@code Clear}. */
    boolean isClear();

    /** Human-facing tag for a hazard, or {@code null} when clear. */
    String tag();

    /** Implicated positions (empty when clear). */
    List<Integer> positions();
  }

  /** No skin-tone / variation-selector abuse present. */
  public record Clear() implements Classification {
    @Override
    public boolean isClear() {
      return true;
    }

    @Override
    public String tag() {
      return null;
    }

    @Override
    public List<Integer> positions() {
      return List.of();
    }
  }

  /**
   * A hazard: the fired sub-threat, the implicated positions, and the
   * (always-empty for this detector) decoded-byte projection — kept for shape
   * parity with the Lean {@code Classification.hazard}.
   */
  public record Hazard(SubThreat sub, List<Integer> positions, List<Integer> decoded)
      implements Classification {
    @Override
    public boolean isClear() {
      return false;
    }

    @Override
    public String tag() {
      return sub.tag();
    }
  }

  /** The structured output of {@link #detect} (mirrors the Lean {@code Verdict}). */
  public record Verdict(
      List<Integer> input,
      Classification classify,
      int skinToneCount,
      int variationSelector15Count,
      int variationSelector16Count) {}

  /**
   * The fixture reason code for a classification, or {@code null} when clear.
   * Mirrors the shared {@code unicode.security.I.skin-tone-variation-forgery.<tag>}
   * scheme keyed by the sub-threat tag.
   */
  public static String reasonCode(Classification classify) {
    String tag = classify.tag();
    return tag == null ? null : "unicode.security.I." + FAMILY + "." + tag;
  }

  // ───────────────────────────────────────────────────────────────────────
  // §2 Emoji property tables (bundled data/emoji-data.txt)
  // ───────────────────────────────────────────────────────────────────────

  private static volatile List<int[]> emojiModifierBaseRanges;
  private static volatile List<int[]> emojiPresentationRanges;

  /**
   * Parse the closed intervals for a single emoji property from emoji-data.txt.
   * Each non-comment row is {@code <range> ; <property> # <comment>}; only rows
   * whose property field is exactly {@code property} are kept. The raw bytes are
   * served by the port's SHA-pinned {@link Security#readResource}.
   */
  private static List<int[]> parseEmojiProperty(String property) {
    List<int[]> out = new ArrayList<>();
    String raw = Security.readResource("emoji-data.txt");
    for (String rawLine : raw.split("\n", -1)) {
      int hash = rawLine.indexOf('#');
      String body = hash >= 0 ? rawLine.substring(0, hash) : rawLine;
      String stripped = body.trim();
      if (stripped.isEmpty()) {
        continue;
      }
      String[] fields = stripped.split(";", -1);
      if (fields.length < 2) {
        continue;
      }
      if (!fields[1].trim().equals(property)) {
        continue;
      }
      String range = fields[0].trim();
      int dots = range.indexOf("..");
      if (dots >= 0) {
        Integer lo = parseHex(range.substring(0, dots).trim());
        Integer hi = parseHex(range.substring(dots + 2).trim());
        if (lo == null || hi == null) {
          continue;
        }
        out.add(new int[] {lo, hi});
      } else {
        Integer single = parseHex(range);
        if (single == null) {
          continue;
        }
        out.add(new int[] {single, single});
      }
    }
    return out;
  }

  private static Integer parseHex(String hex) {
    try {
      return Integer.parseInt(hex, 16);
    } catch (NumberFormatException parseError) {
      return null;
    }
  }

  private static List<int[]> emojiModifierBaseRanges() {
    List<int[]> local = emojiModifierBaseRanges;
    if (local == null) {
      synchronized (SkinToneVariationForgery.class) {
        local = emojiModifierBaseRanges;
        if (local == null) {
          local = parseEmojiProperty("Emoji_Modifier_Base");
          emojiModifierBaseRanges = local;
        }
      }
    }
    return local;
  }

  private static List<int[]> emojiPresentationRanges() {
    List<int[]> local = emojiPresentationRanges;
    if (local == null) {
      synchronized (SkinToneVariationForgery.class) {
        local = emojiPresentationRanges;
        if (local == null) {
          local = parseEmojiProperty("Emoji_Presentation");
          emojiPresentationRanges = local;
        }
      }
    }
    return local;
  }

  private static boolean inRanges(List<int[]> ranges, int cp) {
    for (int[] range : ranges) {
      if (range[0] <= cp && cp <= range[1]) {
        return true;
      }
    }
    return false;
  }

  // ───────────────────────────────────────────────────────────────────────
  // §3 Core predicates
  // ───────────────────────────────────────────────────────────────────────

  /**
   * True iff {@code cp} is an emoji skin-tone modifier. Reuses the port's own
   * {@link EmojiZwjIntegrity#isEmojiModifier} ({@code U+1F3FB..U+1F3FF}).
   */
  public static boolean isSkinTone(int cp) {
    return EmojiZwjIntegrity.isEmojiModifier(cp);
  }

  /** True iff {@code cp} has {@code Emoji_Modifier_Base} per emoji-data.txt. */
  public static boolean isSkinToneBase(int cp) {
    return inRanges(emojiModifierBaseRanges(), cp);
  }

  /** True iff {@code cp} has {@code Emoji_Presentation} per emoji-data.txt. */
  public static boolean isEmojiPresentation(int cp) {
    return inRanges(emojiPresentationRanges(), cp);
  }

  /** True iff {@code cp} is {@code U+FE0E} (VS15, text-style variation selector). */
  public static boolean isVs15(int cp) {
    return cp == VS15;
  }

  /** True iff {@code cp} is {@code U+FE0F} (VS16, emoji-style variation selector). */
  public static boolean isVs16(int cp) {
    return cp == VS16;
  }

  // ───────────────────────────────────────────────────────────────────────
  // §4 Sub-detectors
  // ───────────────────────────────────────────────────────────────────────

  /**
   * First position whose next two codepoints are both skin-tone modifiers, as a
   * {@link StackedSkinTones} with {@code base_pos} and the first two modifiers, or
   * {@code null} when none exists.
   */
  private static StackedSkinTones firstStackedSkinTones(List<Integer> input) {
    for (int i = 0; i < input.size(); i++) {
      if (i + 2 < input.size()) {
        int m1 = input.get(i + 1);
        int m2 = input.get(i + 2);
        if (isSkinTone(m1) && isSkinTone(m2)) {
          List<Integer> modifiers = new ArrayList<>();
          modifiers.add(m1);
          modifiers.add(m2);
          return new StackedSkinTones(i, modifiers);
        }
      }
    }
    return null;
  }

  /**
   * First skin-tone modifier whose preceding codepoint is NOT
   * {@code Emoji_Modifier_Base}, as an {@link InvalidSkinToneTarget}, or
   * {@code null} when none exists.
   */
  private static InvalidSkinToneTarget firstInvalidSkinToneTarget(List<Integer> input) {
    for (int i = 0; i < input.size(); i++) {
      if (i + 1 < input.size()) {
        int cp = input.get(i + 1);
        if (isSkinTone(cp) && !isSkinToneBase(input.get(i))) {
          return new InvalidSkinToneTarget(i, input.get(i), cp);
        }
      }
    }
    return null;
  }

  /**
   * First {@code U+FE0E} whose preceding codepoint has {@code Emoji_Presentation},
   * as a {@link ForcedTextStyle}, or {@code null} when none exists.
   */
  private static ForcedTextStyle firstForcedTextStyle(List<Integer> input) {
    for (int i = 0; i < input.size(); i++) {
      if (i + 1 < input.size()) {
        int cp = input.get(i + 1);
        if (isVs15(cp) && isEmojiPresentation(input.get(i))) {
          return new ForcedTextStyle(i, input.get(i));
        }
      }
    }
    return null;
  }

  private static int skinToneCount(List<Integer> input) {
    int count = 0;
    for (int cp : input) {
      if (isSkinTone(cp)) {
        count++;
      }
    }
    return count;
  }

  private static int vs15Count(List<Integer> input) {
    int count = 0;
    for (int cp : input) {
      if (isVs15(cp)) {
        count++;
      }
    }
    return count;
  }

  private static int vs16Count(List<Integer> input) {
    int count = 0;
    for (int cp : input) {
      if (isVs16(cp)) {
        count++;
      }
    }
    return count;
  }

  // ───────────────────────────────────────────────────────────────────────
  // §5 Top-level detection
  // ───────────────────────────────────────────────────────────────────────

  /** The SkinToneVariationForgery detection function. */
  public static Verdict detect(List<Integer> input) {
    int stc = skinToneCount(input);
    int v15 = vs15Count(input);
    int v16 = vs16Count(input);

    Classification classification;
    StackedSkinTones stacked = firstStackedSkinTones(input);
    if (stacked != null) {
      // Priority 1: a base followed by two stacked skin tones.
      List<Integer> positions = new ArrayList<>();
      for (int i = 0; i < stacked.modifiers().size(); i++) {
        positions.add(stacked.basePos() + 1 + i);
      }
      classification = new Hazard(stacked, positions, List.of());
    } else {
      InvalidSkinToneTarget invalid = firstInvalidSkinToneTarget(input);
      if (invalid != null) {
        // Priority 2: a skin tone on a non-modifier-base.
        List<Integer> positions = new ArrayList<>();
        positions.add(invalid.basePos() + 1);
        classification = new Hazard(invalid, positions, List.of());
      } else {
        ForcedTextStyle forced = firstForcedTextStyle(input);
        if (forced != null) {
          // Priority 3: VS15 forcing text style on an emoji-presentation cp.
          List<Integer> positions = new ArrayList<>();
          positions.add(forced.basePos() + 1);
          classification = new Hazard(forced, positions, List.of());
        } else {
          classification = new Clear();
        }
      }
    }

    return new Verdict(new ArrayList<>(input), classification, stc, v15, v16);
  }
}

File: ports/dotnet/src/UnicodeSecurity/Identity/SkinToneVariationForgery.cs
namespace UnicodeSecurity;

// skin-tone-variation-forgery (I layer) — skin-tone modifier and
// variation-selector abuse on emoji bases per UTS #51 (identity-layer
// detector).
//
// Direct port of the verified Rust reference implementation, itself a
// byte-faithful transliteration of
// Unicode/Security/Identity/SkinToneVariationForgery.lean.
//
// Threat model. Tier A1. An adversary places a skin-tone modifier on a
// codepoint that does NOT bear Emoji_Modifier_Base, stacks multiple skin-tones
// on one base, or forces a text-style render on an emoji-default codepoint via
// U+FE0E (VS15) — sometimes to hide a payload-bearing glyph in plain sight.
//
// Distinct from VariationSelectorPayload (pair-aligned VS runs that decode to
// bytes): this catches the orthogonal case of semantic VS / skin-tone misuse on
// a single base. Both can fire on the same input; SourceDisplayDivergence
// aggregates.
//
// It reuses the port's own emoji property tables (the bundled emoji-data.txt),
// never a host emoji library. The skin-tone modifier predicate reuses the
// port's own EmojiZwjIntegrity.IsEmojiModifier (U+1F3FB..U+1F3FF); the
// Emoji_Modifier_Base and Emoji_Presentation predicates parse the same bundled
// emoji-data.txt for their respective property rows.
//
// Sub-threats (priority order):
//   1. StackedSkinTones      a base immediately followed by >= 2 skin-tone modifiers.
//   2. InvalidSkinToneTarget a skin-tone modifier on a non-Emoji_Modifier_Base.
//   3. ForcedTextStyle       U+FE0E on an Emoji_Presentation codepoint.
public static partial class Security
{
    public static class SkinToneVariationForgery
    {
        // ─────────────────────────────────────────────────────────────────
        // §1 Types
        // ─────────────────────────────────────────────────────────────────

        /// <summary>Sub-threat enumeration for SkinToneVariationForgery, in
        /// priority order. Each variant overrides <see cref="Tag"/>, so the
        /// fixture-row tag is resolved polymorphically with no catch-all.</summary>
        public abstract record SubThreat
        {
            /// <summary>Fixture-row tag string for this sub-threat (matches the
            /// Lean SubThreat.tag).</summary>
            public abstract string Tag { get; }
        }

        /// <summary>A base at <see cref="BasePos"/> followed by >= 2 skin-tone
        /// modifiers (<see cref="Modifiers"/>, the first two stacked
        /// modifiers).</summary>
        public sealed record StackedSkinTones(int BasePos, IReadOnlyList<int> Modifiers) : SubThreat
        {
            public override string Tag => "StackedSkinTones";
        }

        /// <summary>A skin-tone <see cref="ModifierCp"/> at
        /// <see cref="BasePos"/> + 1 on a non-modifier-base
        /// <see cref="BaseCp"/>.</summary>
        public sealed record InvalidSkinToneTarget(int BasePos, int BaseCp, int ModifierCp) : SubThreat
        {
            public override string Tag => "InvalidSkinToneTarget";
        }

        /// <summary>A U+FE0E at <see cref="BasePos"/> + 1 forcing text-style on
        /// an Emoji_Presentation <see cref="BaseCp"/>.</summary>
        public sealed record ForcedTextStyle(int BasePos, int BaseCp) : SubThreat
        {
            public override string Tag => "ForcedTextStyle";
        }

        /// <summary>Top-level classification for SkinToneVariationForgery.</summary>
        public abstract record Classification
        {
            /// <summary>True iff the classification is Clear.</summary>
            public abstract bool IsClear { get; }

            /// <summary>Human-facing tag for a hazard, or null when clear.</summary>
            public abstract string? Tag { get; }

            /// <summary>Implicated positions (empty when clear).</summary>
            public abstract IReadOnlyList<int> Positions { get; }

            /// <summary>Fully-qualified reason code
            /// (unicode.security.I.skin-tone-variation-forgery.&lt;tag&gt;)
            /// routed through the port's shared reason-code machinery, or null
            /// when clear.</summary>
            public string? ReasonCode =>
                Tag is null ? null : Security.ReasonCode(Family.SkinToneVariationForgery, Tag);
        }

        /// <summary>No skin-tone / variation-selector abuse present.</summary>
        public sealed record Clear : Classification
        {
            public override bool IsClear => true;
            public override string? Tag => null;
            public override IReadOnlyList<int> Positions => System.Array.Empty<int>();
        }

        /// <summary>A hazard: the fired sub-threat, the implicated positions, and
        /// the (always-empty for this detector) decoded-byte projection, kept for
        /// shape parity with the Lean Classification.hazard.</summary>
        public sealed record Hazard(
            SubThreat Sub,
            IReadOnlyList<int> HazardPositions,
            IReadOnlyList<int> Decoded) : Classification
        {
            public override bool IsClear => false;
            public override string? Tag => Sub.Tag;
            public override IReadOnlyList<int> Positions => HazardPositions;
        }

        /// <summary>The structured output of <see cref="Detect"/> (mirrors the
        /// Lean Verdict).</summary>
        public sealed record Verdict(
            IReadOnlyList<int> Input,
            Classification Classify,
            int SkinToneCount,
            int VariationSelector15Count,
            int VariationSelector16Count);

        // ─────────────────────────────────────────────────────────────────
        // §2 Core predicates (reuse the port's own emoji tables)
        // ─────────────────────────────────────────────────────────────────

        private static List<(int Lo, int Hi)>? emojiModifierBaseRanges;
        private static List<(int Lo, int Hi)>? emojiPresentationRanges;

        /// <summary>Parse the closed intervals for a single emoji property from
        /// emoji-data.txt. Each non-comment row is
        /// <c>&lt;range&gt; ; &lt;property&gt; # &lt;comment&gt;</c>; keep only
        /// rows whose property field matches <paramref name="property"/>
        /// exactly. Mirrors the AiWatermarkDetectability Emoji-row parser,
        /// parameterised on the property name.</summary>
        private static List<(int Lo, int Hi)> ParseEmojiProperty(string property, string raw)
        {
            var output = new List<(int, int)>();
            foreach (var rawLine in raw.Split('\n'))
            {
                var hash = rawLine.IndexOf('#');
                var body = (hash >= 0 ? rawLine.Substring(0, hash) : rawLine).Trim();
                if (body.Length == 0) continue;
                var fields = body.Split(';');
                if (fields.Length < 2) continue;
                if (fields[1].Trim() != property) continue;
                var range = fields[0].Trim();
                var dots = range.IndexOf("..", StringComparison.Ordinal);
                if (dots >= 0)
                {
                    var lo = Security.ParseHex(range.Substring(0, dots));
                    var hi = Security.ParseHex(range.Substring(dots + 2));
                    if (lo is not null && hi is not null) output.Add((lo.Value, hi.Value));
                }
                else
                {
                    var single = Security.ParseHex(range);
                    if (single is not null) output.Add((single.Value, single.Value));
                }
            }
            return output;
        }

        private static List<(int Lo, int Hi)> EmojiModifierBaseRanges()
        {
            emojiModifierBaseRanges ??=
                ParseEmojiProperty("Emoji_Modifier_Base", Security.ReadDataFile("emoji-data.txt"));
            return emojiModifierBaseRanges;
        }

        private static List<(int Lo, int Hi)> EmojiPresentationRanges()
        {
            emojiPresentationRanges ??=
                ParseEmojiProperty("Emoji_Presentation", Security.ReadDataFile("emoji-data.txt"));
            return emojiPresentationRanges;
        }

        /// <summary>True iff <paramref name="cp"/> is an emoji skin-tone modifier.
        /// Reuses the port's own predicate (U+1F3FB..U+1F3FF), never a host
        /// emoji library.</summary>
        public static bool IsSkinTone(int cp) => EmojiZwjIntegrity.IsEmojiModifier(cp);

        /// <summary>True iff <paramref name="cp"/> has Emoji_Modifier_Base per
        /// emoji-data.txt.</summary>
        public static bool IsSkinToneBase(int cp)
        {
            foreach (var (lo, hi) in EmojiModifierBaseRanges())
            {
                if (lo <= cp && cp <= hi) return true;
            }
            return false;
        }

        /// <summary>True iff <paramref name="cp"/> has Emoji_Presentation per
        /// emoji-data.txt.</summary>
        public static bool IsEmojiPresentation(int cp)
        {
            foreach (var (lo, hi) in EmojiPresentationRanges())
            {
                if (lo <= cp && cp <= hi) return true;
            }
            return false;
        }

        /// <summary>True iff <paramref name="cp"/> is U+FE0E (VS15, text-style
        /// variation selector).</summary>
        public static bool IsVs15(int cp) => cp == 0xFE0E;

        /// <summary>True iff <paramref name="cp"/> is U+FE0F (VS16, emoji-style
        /// variation selector).</summary>
        public static bool IsVs16(int cp) => cp == 0xFE0F;

        // ─────────────────────────────────────────────────────────────────
        // §3 Sub-detectors
        // ─────────────────────────────────────────────────────────────────

        /// <summary>First position <c>i</c> whose next two codepoints are both
        /// skin-tone modifiers, as (basePos, [mod1, mod2]). Null when
        /// none.</summary>
        private static (int BasePos, List<int> Modifiers)? FirstStackedSkinTones(IReadOnlyList<int> input)
        {
            for (var i = 0; i < input.Count; i++)
            {
                if (i + 2 < input.Count && IsSkinTone(input[i + 1]) && IsSkinTone(input[i + 2]))
                {
                    return (i, new List<int> { input[i + 1], input[i + 2] });
                }
            }
            return null;
        }

        /// <summary>First skin-tone modifier whose preceding codepoint is NOT
        /// Emoji_Modifier_Base, as (basePos, baseCp, modifierCp). Null when
        /// none.</summary>
        private static (int BasePos, int BaseCp, int ModifierCp)? FirstInvalidSkinToneTarget(IReadOnlyList<int> input)
        {
            for (var i = 0; i < input.Count; i++)
            {
                if (i + 1 < input.Count && IsSkinTone(input[i + 1]) && !IsSkinToneBase(input[i]))
                {
                    return (i, input[i], input[i + 1]);
                }
            }
            return null;
        }

        /// <summary>First U+FE0E whose preceding codepoint has
        /// Emoji_Presentation, as (basePos, baseCp). Null when none.</summary>
        private static (int BasePos, int BaseCp)? FirstForcedTextStyle(IReadOnlyList<int> input)
        {
            for (var i = 0; i < input.Count; i++)
            {
                if (i + 1 < input.Count && IsVs15(input[i + 1]) && IsEmojiPresentation(input[i]))
                {
                    return (i, input[i]);
                }
            }
            return null;
        }

        private static int SkinToneCountOf(IReadOnlyList<int> input)
        {
            var count = 0;
            foreach (var cp in input)
            {
                if (IsSkinTone(cp)) count++;
            }
            return count;
        }

        private static int Vs15Count(IReadOnlyList<int> input)
        {
            var count = 0;
            foreach (var cp in input)
            {
                if (IsVs15(cp)) count++;
            }
            return count;
        }

        private static int Vs16Count(IReadOnlyList<int> input)
        {
            var count = 0;
            foreach (var cp in input)
            {
                if (IsVs16(cp)) count++;
            }
            return count;
        }

        // ─────────────────────────────────────────────────────────────────
        // §4 Top-level detection
        // ─────────────────────────────────────────────────────────────────

        /// <summary>The SkinToneVariationForgery detection function.</summary>
        public static Verdict Detect(IReadOnlyList<int> input)
        {
            var stc = SkinToneCountOf(input);
            var v15 = Vs15Count(input);
            var v16 = Vs16Count(input);

            Classification classification;
            if (FirstStackedSkinTones(input) is (int stackedBasePos, List<int> modifiers))
            {
                // Priority 1: a base followed by two stacked skin tones.
                var positions = new List<int>();
                for (var i = 0; i < modifiers.Count; i++)
                {
                    positions.Add(stackedBasePos + 1 + i);
                }
                classification = new Hazard(
                    new StackedSkinTones(stackedBasePos, modifiers),
                    positions,
                    new List<int>());
            }
            else if (FirstInvalidSkinToneTarget(input) is (int invalidBasePos, int baseCp, int modifierCp))
            {
                // Priority 2: a skin tone on a non-modifier-base.
                classification = new Hazard(
                    new InvalidSkinToneTarget(invalidBasePos, baseCp, modifierCp),
                    new List<int> { invalidBasePos + 1 },
                    new List<int>());
            }
            else if (FirstForcedTextStyle(input) is (int forcedBasePos, int forcedBaseCp))
            {
                // Priority 3: VS15 forcing text style on an emoji-presentation cp.
                classification = new Hazard(
                    new ForcedTextStyle(forcedBasePos, forcedBaseCp),
                    new List<int> { forcedBasePos + 1 },
                    new List<int>());
            }
            else
            {
                classification = new Clear();
            }

            return new Verdict(input.ToList(), classification, stc, v15, v16);
        }
    }
}
